### PR TITLE
Pebble integration

### PR DESCRIPTION
**Pull request by vsmid**
This is a Pebble template engine extension integration pull request. 
I think it should be worth having this in Ozark's arsenal :-)

### CODE REVIEW

File: ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleConfigurationProducer.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  public Properties loadFromFile(String fileName) {
+    Properties props = new Properties();
+
+    try (InputStream config = Thread.currentThread().getContextClassLoader().getResourceAsStream(fileName)) {
+      if (Objects.nonNull(config)) {
+        props.load(config);
+      }
+    } catch (IOException e) {
```
Comment (sputnikci) [inline ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleConfigurationProducer.java]: \[Checkstyle\] ERROR: Must have at least one statement\.

File: ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleEngineProducer.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  private Properties pebbleConfiguration;
+  private ServletContext servletContext;
+
+  @Inject
+  public PebbleEngineProducer(Properties pebbleConfiguration, ServletContext servletContext) {
+    this.pebbleConfiguration = pebbleConfiguration;
+    this.servletContext = servletContext;
+  }
+
+  @Produces
```
Comment (sputnikci) [inline ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleEngineProducer.java]: \[Checkstyle\] ERROR: NCSS for this method is 54 \(max allowed is 50\)\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public class CustomExecutorService implements ExecutorService {
+
+  @Override
+  public void shutdown() {
+    throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java]: \[Checkstyle\] ERROR: Don't use trailing comments\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  @Override
+  public List<Runnable> shutdownNow() {
+    throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java]: \[Checkstyle\] ERROR: Don't use trailing comments\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  @Override
+  public boolean isShutdown() {
+    throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java]: \[Checkstyle\] ERROR: Don't use trailing comments\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  @Override
+  public boolean isTerminated() {
+    throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java]: \[Checkstyle\] ERROR: Don't use trailing comments\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  @Override
+  public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+    throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java]: \[Checkstyle\] ERROR: Don't use trailing comments\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  @Override
+  public Future<?> submit(Runnable task) {
+    throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
+  }
+
+  @Override
+  public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks) throws InterruptedException {
+    throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java]: \[Checkstyle\] ERROR: Don't use trailing comments\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  @Override
+  public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit) throws InterruptedException {
+    throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java]: \[Checkstyle\] ERROR: Don't use trailing comments\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  @Override
+  public <T> T invokeAny(Collection<? extends Callable<T>> tasks) throws InterruptedException, ExecutionException {
+    throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java]: \[Checkstyle\] ERROR: Don't use trailing comments\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  @Override
+  public <T> T invokeAny(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+    throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java]: \[Checkstyle\] ERROR: Don't use trailing comments\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  @Override
+  public void execute(Runnable command) {
+    throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/CustomExecutorService.java]: \[Checkstyle\] ERROR: Don't use trailing comments\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/PebbleController.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+import javax.mvc.annotation.View;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+
+@Path("")
+@Controller
+public class PebbleController {
+
+  @Inject
+  Models models;
```
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/PebbleController.java]: \[Checkstyle\] ERROR: Variable 'models' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleConfigurationProducerTest.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.Properties;
+import org.glassfish.ozark.ext.pebble.PebbleConfigurationProducer;
+import org.glassfish.ozark.ext.pebble.PebbleProperty;
+import static org.junit.Assert.*;
+import org.junit.Before;
+import org.junit.Test;
+
+public class PebbleConfigurationProducerTest {
+
+  PebbleConfigurationProducer pebbleConfigurationProducer;
```
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleConfigurationProducerTest.java]: \[Checkstyle\] ERROR: Variable 'pebbleConfigurationProducer' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.glassfish.ozark.test.pebble;
+
+import com.google.common.cache.Cache;
+import com.mitchellbosecke.pebble.PebbleEngine;
+import com.mitchellbosecke.pebble.cache.BaseTagCacheKey;
+import com.mitchellbosecke.pebble.extension.NodeVisitor;
```
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java]: \[Checkstyle\] ERROR: Unused import \- com\.mitchellbosecke\.pebble\.extension\.NodeVisitor\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+import java.util.stream.IntStream;
+import org.glassfish.ozark.ext.pebble.PebbleEngineProducer;
+import org.glassfish.ozark.ext.pebble.PebbleProperty;
+import static org.junit.Assert.*;
+import org.junit.Before;
+import org.junit.Test;
+
+public class PebbleEngineProducerTest {
+
+  Properties properties;
```
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java]: \[Checkstyle\] ERROR: Variable 'properties' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+  PebbleEngineProducer pebbleEngineProducer;
```
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java]: \[Checkstyle\] ERROR: Variable 'pebbleEngineProducer' must be private and have accessor methods\.

**Comment by chkal** [comment]
Hey. That's awesome. I'll have a detailed look tomorrow. 

Please ignore the Checkstyle stuff for now. Not sure if we want to keep it. 

But the tests also broke. Could you have a look?

**Comment by vsmid** [comment]
Hi, 

yeah, sure, looking at it :-)

tests broke due to missing ozark dependency for test scope i think.

**Comment by vsmid** [comment]
i have pushed some updates,  how can i start the build again?

File: ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleConfigurationProducer.java (hunk 1 above)
Comment (sputnikci) [inline ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleConfigurationProducer.java]: \[Checkstyle\] ERROR: Must have at least one statement\.

File: ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleEngineProducer.java (hunk 2 above)
Comment (sputnikci) [inline ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleEngineProducer.java]: \[Checkstyle\] ERROR: NCSS for this method is 54 \(max allowed is 50\)\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/PebbleController.java (hunk 13 above)
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/PebbleController.java]: \[Checkstyle\] ERROR: Variable 'models' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleConfigurationProducerTest.java (hunk 14 above)
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleConfigurationProducerTest.java]: \[Checkstyle\] ERROR: Variable 'pebbleConfigurationProducer' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+import java.util.stream.IntStream;
+import org.glassfish.ozark.ext.pebble.PebbleEngineProducer;
+import org.glassfish.ozark.ext.pebble.PebbleProperty;
+import static org.junit.Assert.*;
+import org.junit.Before;
+import org.junit.Test;
+
+public class PebbleEngineProducerTest {
+
+  Properties properties;
```
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java]: \[Checkstyle\] ERROR: Variable 'properties' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java [hunk 19]
```diff
@@ -0,0 +1,152 @@
...
+  PebbleEngineProducer pebbleEngineProducer;
```
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java]: \[Checkstyle\] ERROR: Variable 'pebbleEngineProducer' must be private and have accessor methods\.

File: ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleConfigurationProducer.java (hunk 1 above)
Comment (sputnikci) [inline ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleConfigurationProducer.java]: \[Checkstyle\] ERROR: Must have at least one statement\.

File: ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleEngineProducer.java (hunk 2 above)
Comment (sputnikci) [inline ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleEngineProducer.java]: \[Checkstyle\] ERROR: NCSS for this method is 54 \(max allowed is 50\)\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/PebbleController.java (hunk 13 above)
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/PebbleController.java]: \[Checkstyle\] ERROR: Variable 'models' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleConfigurationProducerTest.java (hunk 14 above)
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleConfigurationProducerTest.java]: \[Checkstyle\] ERROR: Variable 'pebbleConfigurationProducer' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java (hunk 18 above)
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java]: \[Checkstyle\] ERROR: Variable 'properties' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java (hunk 19 above)
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java]: \[Checkstyle\] ERROR: Variable 'pebbleEngineProducer' must be private and have accessor methods\.

File: ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleConfigurationProducer.java (hunk 1 above)
Comment (sputnikci) [inline ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleConfigurationProducer.java]: \[Checkstyle\] ERROR: Must have at least one statement\.

File: ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleEngineProducer.java (hunk 2 above)
Comment (sputnikci) [inline ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleEngineProducer.java]: \[Checkstyle\] ERROR: NCSS for this method is 54 \(max allowed is 50\)\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/PebbleController.java (hunk 13 above)
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/PebbleController.java]: \[Checkstyle\] ERROR: Variable 'models' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleConfigurationProducerTest.java (hunk 14 above)
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleConfigurationProducerTest.java]: \[Checkstyle\] ERROR: Variable 'pebbleConfigurationProducer' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java (hunk 18 above)
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java]: \[Checkstyle\] ERROR: Variable 'properties' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java (hunk 19 above)
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java]: \[Checkstyle\] ERROR: Variable 'pebbleEngineProducer' must be private and have accessor methods\.

File: ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleConfigurationProducer.java (hunk 1 above)
Comment (sputnikci) [inline ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleConfigurationProducer.java]: \[Checkstyle\] ERROR: Must have at least one statement\.

File: ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleEngineProducer.java (hunk 2 above)
Comment (sputnikci) [inline ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleEngineProducer.java]: \[Checkstyle\] ERROR: NCSS for this method is 54 \(max allowed is 50\)\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/PebbleController.java (hunk 13 above)
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/PebbleController.java]: \[Checkstyle\] ERROR: Variable 'models' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleConfigurationProducerTest.java (hunk 14 above)
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleConfigurationProducerTest.java]: \[Checkstyle\] ERROR: Variable 'pebbleConfigurationProducer' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java (hunk 18 above)
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java]: \[Checkstyle\] ERROR: Variable 'properties' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java (hunk 19 above)
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java]: \[Checkstyle\] ERROR: Variable 'pebbleEngineProducer' must be private and have accessor methods\.

File: ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleConfigurationProducer.java (hunk 1 above)
Comment (sputnikci) [inline ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleConfigurationProducer.java]: \[Checkstyle\] ERROR: Must have at least one statement\.

File: ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleEngineProducer.java (hunk 2 above)
Comment (sputnikci) [inline ext/pebble/src/main/java/org/glassfish/ozark/ext/pebble/PebbleEngineProducer.java]: \[Checkstyle\] ERROR: NCSS for this method is 54 \(max allowed is 50\)\.

File: test/pebble/src/main/java/org/glassfish/ozark/test/pebble/PebbleController.java (hunk 13 above)
Comment (sputnikci) [inline test/pebble/src/main/java/org/glassfish/ozark/test/pebble/PebbleController.java]: \[Checkstyle\] ERROR: Variable 'models' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleConfigurationProducerTest.java (hunk 14 above)
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleConfigurationProducerTest.java]: \[Checkstyle\] ERROR: Variable 'pebbleConfigurationProducer' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java (hunk 18 above)
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java]: \[Checkstyle\] ERROR: Variable 'properties' must be private and have accessor methods\.

File: test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java (hunk 19 above)
Comment (sputnikci) [inline test/pebble/src/test/java/org/glassfish/ozark/test/pebble/PebbleEngineProducerTest.java]: \[Checkstyle\] ERROR: Variable 'pebbleEngineProducer' must be private and have accessor methods\.

**Comment by vsmid** [comment]
Ok, tests are now passing. It turns out that Glassfish bundled version needed Ozark so i put it as compile scoped dependency for test project.

**Comment by chkal** [comment]
Ok, great! I had a quick look at the code and it looks great! Well structured and with a lot of tests. So merging this right away. :beers:

**Comment by vsmid** [comment]
thanks a lot!😊

**Pull request merged**